REGRESSION: sreality.cz: pinch zooming on maps causes tab to enter Tab Overview instead of interacting with map

rr-codes opened the pull request:
#### 887ab6d07a2756502d80ca21aa9eec1a0b2f7a76
<pre>
REGRESSION: sreality.cz: pinch zooming on maps causes tab to enter Tab Overview instead of interacting with map
<a href="https://bugs.webkit.org/show_bug.cgi?id=252023">https://bugs.webkit.org/show_bug.cgi?id=252023</a>
rdar://104329203

Reviewed by Wenson Hsieh.

<a href="https://github.com/WebKit/WebKit/commit/b7a998fb59f59222ea76fe60141f38cfd2fc8cde">https://github.com/WebKit/WebKit/commit/b7a998fb59f59222ea76fe60141f38cfd2fc8cde</a> introduced a new deferring gesture recognizer
for pan and pinch gestures. This had the conseequence of now letting Safari take
precedence over consuming these gestures rather than WebKit, causing the tab overview
to appear instead of zooming out of the map.

This PR adds a new private delegate method to `WKUIDelegate` that allows clients to specify
if a recognizer can be prevented by touch events. Then, if the gesture is also one that is
installed over the web view, the web view will take precedence.

* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _gestureRecognizerCanBePreventedByTouchEvents:]):
(-[WKContentView deferringGestureRecognizer:shouldDeferOtherGestureRecognizer:]):

Canonical link: <a href="https://commits.webkit.org/260294@main">https://commits.webkit.org/260294@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/85bb734fd41c09640d705911c54a7eeaf97a6a7b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107824 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16878 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40710 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116962 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116333 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18280 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8193 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99999 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113581 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13799 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97006 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41498 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95706 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28647 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83342 "Build is in progress. Recent messages:Running worker_preparation") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9828 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/29995 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10523 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6890 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15981 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49579 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7121 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12090 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -1763,6 +1763,18 @@ - (BOOL)_touchEventsMustRequireGestureRecognizerToFail:(UIGestureRecognizer *)ge
+{
+    auto uiDelegate = static_cast<id<WKUIDelegatePrivate>>(self.webView.UIDelegate);
+    if (!uiDelegate)
+        return NO;
```
This check is unnecessary. If `uiDelegate` is `nil`, the `respondsToSelector` check below will fail.

The method could just `return [uiDelegate respondsToSelector:@selector(_webView:gestureRecognizerCanBePreventedByTouchEvents:)] && [uiDelegate _webView:self.webView gestureRecognizerCanBePreventedByTouchEvents:gestureRecognizer]`.

rr-codes committed:
REGRESSION: sreality.cz: pinch zooming on maps causes tab to enter Tab Overview instead of interacting with map
https://bugs.webkit.org/show_bug.cgi?id=252023
rdar://104329203

Reviewed by Wenson Hsieh.

https://github.com/WebKit/WebKit/commit/b7a998fb59f59222ea76fe60141f38cfd2fc8cde introduced a new deferring gesture recognizer
for pan and pinch gestures. This had the conseequence of now letting Safari take
precedence over consuming these gestures rather than WebKit, causing the tab overview
to appear instead of zooming out of the map.

This PR adds a new private delegate method to `WKUIDelegate` that allows clients to specify
if a recognizer can be prevented by touch events. Then, if the gesture is also one that is
installed over the web view, the web view will take precedence.

* Source/WebKit/UIProcess/API/Cocoa/WKUIDelegatePrivate.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _gestureRecognizerCanBePreventedByTouchEvents:]):
(-[WKContentView deferringGestureRecognizer:shouldDeferOtherGestureRecognizer:]):

Canonical link: https://commits.webkit.org/260294@main
